### ISSUE TITLE
ImportError -- not handing relative imports?

### ISSUE THREAD

**Issue by adammenges** [pyinstaller/pyinstaller#1430]
I'm trying to use pyinstaller on a relatively simple script that uses scikit-learn, however I get the following:

```
Traceback (most recent call last):
  File "<string>", line 6, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/classification", line 4, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.feature_extraction", line 7, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.feature_extraction.dict_vectorizer", line 15, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils", line 16, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils.class_weight", line 8, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils.fixes", line 318, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.sparse.linalg", line 109, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.sparse.linalg.isolve", line 8, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.sparse.linalg.isolve.lgmres", line 8, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.linalg", line 188, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 409, in load_module
    module = imp.load_module(fullname, fp, filename, self._c_ext_tuple)
  File "scipy/linalg/_decomp_update.pyx", line 1, in init scipy.linalg._decomp_update (/private/var/folders/1q/gffnf5w53mv545vj96x72fzr0000gn/T/pip-build-EPg6Ph/scipy-dev/scipy/linalg/_decomp_update.c:35835)
ImportError: No module named cython_blas
```

Importing:

```python
import pickle, requests, json, os, itertools, time, string, sklearn, re, nltk, math

from textblob import TextBlob
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.svm import SVC, LinearSVC
from sklearn.metrics import classification_report, f1_score, accuracy_score, confusion_matrix
from sklearn.pipeline import Pipeline
from sklearn.grid_search import GridSearchCV
from sklearn.cross_validation import StratifiedKFold, cross_val_score, train_test_split
from nltk.corpus import stopwords

from collections import Counter
```

**Comment by codewarrior0** [comment]
We will need to add a hook file for the module `sklearn`. It probably just needs to list some hiddenimports, similar to the [skimage.transform hook](https://github.com/pyinstaller/pyinstaller/blob/python3/PyInstaller/hooks/hook-skimage.transform.py)

In the meantime, could you try adding `scipy.linalg.cython_blas` to your `hiddenimports`?

**Comment by adammenges** [comment]
Thanks @codewarrior0 -- got that guy working, now I'm on:

```
Traceback (most recent call last):
  File "<string>", line 9, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/classification", line 4, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.feature_extraction", line 7, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.feature_extraction.dict_vectorizer", line 15, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils", line 16, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils.class_weight", line 8, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/sklearn.utils.fixes", line 318, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.sparse.linalg", line 113, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.sparse.linalg.matfuncs", line 19, in <module>
  File "/usr/local/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.misc", line 46, in <module>
ImportError: cannot import name doccer
```

Figured I could solve it the same way with a hookfile `hook-scipy.misc.py`:
```python
hiddenimports = ['doccer']
```

But it doesn't this time.

How should I be interpreting that output to know which module to add my hidden imports to?

**Comment by adammenges** [comment]
Also, is there a quicker way to go through this, rather then trial and error? I see the docs mention `python -v`, but based on that output I'm not sure which imports are hidden and which are not. 

Appreciate the help.

**Comment by adammenges** [comment]
Oh, and once I'm done going through all these scipy guys, I'll submit a pull request.

**Comment by matysek** [comment]
```
File "/Users/adammenges/Development/llks/build/llks/out00-PYZ.pyz/scipy.misc", line 46, in <module>
ImportError: cannot import name doccer
```
@adammenges It looks like hiddenimport won't help you this time. This is more some kind of syntax issue or anything like that. The error complains just that the module cannot be imported but not that the module 'is missingr' or 'could not be found'.

**Comment by adammenges** [comment]
Hum, why would this happen?

For what its worth, perhaps this is obvious, but the script runs without a hiccup. I only see this on the pyinstaller version.

# Sent from my iPhone, excuse my brevity.

**Comment by codewarrior0** [comment]
`scipy.misc.doccer` should be getting found normally. The syntax it uses is `from . import doccer` so I'm wondering if something is broken with the way we handle relative imports.

**Comment by matysek** [comment]
@codewarrior0 relative imports seem working. I moved a few tests to py.test and they are passing. So maybe a new case with relative imports that we do not handle properly?

@adammenges What is your exact python version?  2.7.10?

**Comment by adammenges** [comment]
@matysek Yep, 2.7.10.

**Comment by adammenges** [comment]
Any updates here? Anything I can do to help?

**Comment by codewarrior0** [comment]
You could try adding `scipy.misc.doccer` to your hiddenimports, but it *really should* be detected normally. We're still looking into this.

**Comment by adammenges** [comment]
Something different then the hook file I mentioned above? I also tried `--hidden-import="scipy.misc.doccer"`, but with no avail.

**Comment by adammenges** [comment]
@matysek For shits and giggles, I checked out e01229a

```bash
⇒  pyinstaller llks.py --additional-hooks-dir=./hooks --clean
14 INFO: PyInstaller: 3.0dev-e01229a
14 INFO: Python: 2.7.10
18 INFO: Platform: Darwin-14.3.0-x86_64-i386-64bit
...
```
(Before I did that I was on pyinstaller 2.1)

And, not sure if you'd expect this as it's a dev branch, but for what it's worth now I see:

```bash
⇒  ./dist/llks/llks -h
mod is NULL - pyimod02_archiveTraceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/PyInstaller-3.0dev_e01229a-py2.7.egg/PyInstaller/loader/pyimod02_archive.py", line 30, in <module>
    import zlib
ImportError: No module named zlib
mod is NULL - pyimod03_importersTraceback (most recent call last):
  File "/usr/local/lib/python2.7/site-packages/PyInstaller-3.0dev_e01229a-py2.7.egg/PyInstaller/loader/pyimod03_importers.py", line 26, in <module>
    from pyimod02_archive import ArchiveReadError, ZlibArchiveReader
ImportError: No module named pyimod02_archive
Traceback (most recent call last):
  File "<string>", line 15, in <module>
ImportError: No module named pyimod03_importers
pyiboot01_bootstrap returned -1
```

Just let me know if there's anything I can do to help reproduce this guy.

**Comment by codewarrior0** [comment]
Would you mind running `(cd bootloader; python waf all)` and try the build again?

**Comment by adammenges** [comment]
Done and done: here's the full session: https://gist.github.com/adammenges/1f8fc820d6d61c692e0f

**Comment by codewarrior0** [comment]
Whoops! Module `zlib` wasn't being included if it isn't a built-in module. Should be fixed in 9e7d883a5cb0e528f39cb8afeb898c422d2fb337, if you'd care to try again.

**Comment by adammenges** [comment]
Sweet, went again, something slightly different this time: https://gist.github.com/adammenges/1f8fc820d6d61c692e0f

**Comment by adammenges** [comment]
Looks like the most important part in there is:

```bash
Security-Alert: try to store file outside of dist-directory. Aborting. '/usr/lib/libobjc.dylib'
```

**Comment by codewarrior0** [comment]
> `Security-Alert: try to store file outside of dist-directory. Aborting. '/usr/lib/libobjc.dylib'`

That's alarming. It looks like it's aborting during the last build step, so you get an incomplete output folder. 

The error means it's using an absolute path where only relative paths are allowed - probably somebody switched the "internal name" and "filesystem name" when adding a file to a TOC somewhere. I can't imagine where, though.

**Comment by codewarrior0** [comment]
Actually, I can imagine where. There was recently some work done on the way binaries used by calls to `ctypes` are found, and I'll bet it's either using a full path for the "internal name", or not excluding binaries that are in the excludes list (which would match anything under `/usr`), or both.

@matysek, you were working on the ctypes stuff, mind taking a look?

**Comment by codewarrior0** [comment]
@adammenges By the way, which Python distribution is that? It doesn't look like the system-provided one.

**Comment by adammenges** [comment]
Right, it's brew's python. The system python is pretty far behind 2.7.10, I believe.

**Comment by codewarrior0** [comment]
Thanks. I can't think of anything else for you to try at the moment, but could you post the file `build/out00-COLLECT.toc` to a gist?

**Comment by leycec** [comment]
This probably has utterly no relevance, but PyInstaller's use of `modulegraph` is basically broken under Python 2.7 at the moment. The reason why is the `ModuleGraph._import_hook()` method, whose behaviour is non-deterministic when the `level` keyword argument is _not_ explicitly passed. Of course, PyInstaller calls this method without explicitly passing `level`.

Specifically:

* Under Python 3, hidden imports are currently imported as **absolute imports.** That's a good thing.
* Under Python 2, hidden imports are currently imported as both absolute imports _and_ implicit relative imports. That's bad, but it gets worse. Thanks to `ModuleGraph` design, Python 2 hidden imports are imported relative to the name of the currently hooked module. Hence, a hook `hook-foo.bar` containing a global `hiddenimports = ['weea.boo']` would attempt to import a package `foo.bar.weea.boo` rather than the desired `weea.boo`.

So, hidden imports are basically broken under Python 2 at the moment. I'll be submitting a fix shortly.

That said, this may have nothing to do with the current issue. _Carry on, code soldiers._

**Comment by adammenges** [comment]
@codewarrior0: `build/out00-COLLECT.toc` doesn't exist. Nor does `build/<project>/out00-COLLECT.toc`

I've only got:
```
build/<projet>/out00-Analysis.toc
build/<projet>/out00-EXE.toc
build/<projet>/out00-PKG.pkg
build/<projet>/out00-PKG.toc
build/<projet>/out00-PYZ.pyz
build/<projet>/out00-PYZ.toc
build/<projet>/out00-Tree.toc
```

**Comment by codewarrior0** [comment]
That's too bad. It aborts before writing `out00-COLLECT.toc`.

`out00-Analysis.toc` might be helpful.

**Comment by codewarrior0** [comment]
Right about now I'm wishing there were a fourth field in a TOC entry that explains why, how, and where the file was added to the TOC. You know, for debugging things like this.

**Comment by matysek** [comment]
> @codewarrior0: build/out00-COLLECT.toc doesn't exist. Nor does build/<project>/out00-COLLECT.toc

@adammenges  This means that you uses onefile mode. Please post `out00-PKG.toc` or `out00-EXE.toc`.

**Comment by codewarrior0** [comment]
@matysek On the contrary, his build logs have these lines:

```
60429 INFO: Building COLLECT out00-COLLECT.toc
Security-Alert: try to store file outside of dist-directory. Aborting. '/usr/lib/libobjc.dylib'
```

**Comment by matysek** [comment]
@codewarrior0 yes, you are right. I didn't see the above comments :-)

@adammenges What is the content of your ` --additional-hooks-dir=./hooks`? Any custom hook could have any attribute wrong.

**Comment by adammenges** [comment]
@matysek:

```bash
⇒  for fp in ./hooks/*.py
do
    echo $fp; cat $fp; echo
done
./hooks/hook-scipy.linalg.py
hiddenimports = ['cython_blas', 'cython_lapack']
./hooks/hook-scipy.misc.py
hiddenimports = ['doccer']
```

**Comment by adammenges** [comment]
@codewarrior0: https://gist.github.com/adammenges/3e4f33f2f0daaebac6eb

**Comment by codewarrior0** [comment]
[analysis TOC, pprinted](https://gist.github.com/codewarrior0/7582176cb61618dfdf0c#file-analysis_pprinted-L7865)

Context of libobjc entry:

```
('random',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/random.py',
   'PYMODULE'),
  ('_strptime',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/_strptime.py',
   'PYMODULE'),
  ('webbrowser',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/webbrowser.py',
   'PYMODULE')],
 [('Python',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/Python',
   'BINARY'),
  ('/usr/lib/libobjc.dylib', '/usr/lib/libobjc.dylib', 'BINARY'),  # <<<<<<<<<<
  ('fcntl',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload/fcntl.so',
   'EXTENSION'),
  ('cStringIO',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload/cStringIO.so',
   'EXTENSION'),
  ('_io',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload/_io.so',
   'EXTENSION'),
  ('zlib',
   '/usr/local/Cellar/python/2.7.10_2/Frameworks/Python.framework/Versions/2.7/lib/python2.7/lib-dynload/zlib.so',
   'EXTENSION'),
```

Not sure what to make of this at the moment.

**Comment by htgoebel** [comment]
Is this solved already? See https://github.com/pyinstaller/pyinstaller/issues/1382#issuecomment-131452540

**Comment by adammenges** [comment]
@htgoebel: I haven't gotten it working, but I'm more then willing to checkout a branch/commit/tag and give it a go. Which one would you like me to try?

**Comment by htgoebel** [comment]
@adammenges I would appreciate if you would test this ASAP. The current python3 branch head is the one to check out.

**Comment by adammenges** [comment]
@htgoebel unfortunately it ends in the same way: `Security-Alert: try to store file outside of dist-directory. Aborting. '/usr/lib/libobjc.dylib'`

https://gist.github.com/adammenges/3cc0e4cade2b517b7371

**Comment by htgoebel** [comment]
@adammenges IC. This is a different issue, see #1466, blocking this one.

**Comment by adammenges** [comment]
@htgoebel Cheers

**Comment by matysek** [comment]
@adammenges Could you try again latest develop if anything was improved since then?

**Comment by adammenges** [comment]
Awesome, seems we're past the `Security-Alert` :smile: 

Here's the output: https://gist.github.com/adammenges/033426e741c91798b2ad

Looks like it's back to failing on a relative import.

**Comment by adammenges** [comment]
cc @matysek & #1466

**Comment by htgoebel** [comment]
Have you verified `libsvm` is included in the frozen app? Use `archive-viewer` for checking.

**Comment by matysek** [comment]
@adammenges Does the title `not handing relative imports` mean relative module names in `hiddenimports` in hooks?
```python
./hooks/hook-scipy.linalg.py
hiddenimports = ['cython_blas', 'cython_lapack']
./hooks/hook-scipy.misc.py
hiddenimports = ['doccer']
```
Please try to use full module names in these hooks.

**Comment by adammenges** [comment]
@htgoebel Not sure if I'm using the tool correctly, but from what I gather in the docs:

`pyi-archive_viewer ./build/llks/out00-PYZ.pyz` prints out a large list of modules, but libsvm isn't one of them.

-------------

@matysek That title was changed up there ^^ when @codewarrior0 had a guess as to what the problem is. Here's the ending snippet from what's posted above: 

```bash
...
  File "/usr/local/lib/python2.7/site-packages/sklearn/svm/base.py", line 8, in <module>
    from . import libsvm, liblinear
ImportError: cannot import name libsvm
llks returned -1
```

The syntax `from . import libsvm` should be covered by pyinstaller, correct?

It's a little interesting that the last file, `"/usr/local/lib/python2.7/site-packages/sklearn/svm/base.py"`, isn't in the dist directory, is that normal?

WRT the `hiddenimports` I'm not sure what you mean by full module names, `scipy.linalg.cython_blas`? Adding what's there now removes an error about `cython_blas` not being available. Is there a way to add `libsvm` to the hidden imports? Forcing it to be available later.

**Comment by matysek** [comment]
@htgoebel `libsvm` - it is probably related to #1555

**Comment by matysek** [comment]
@adammenges  Try this version, if that fixes it: 825ff6868cf949e99ef08f2b3230c5b08a268bb0

**Comment by htgoebel** [comment]
@adammenges 
> prints out a large list of modules, but libsvm isn't one of them.

Well, this means the file is not packed into the PYZ. But as weknow now, i's an extension-module, so it would not be in the PYZ anyway. But it is missing, see #1555.
